### ISSUE TITLE
How to change clock_period in sim_input.yml

### ISSUE THREAD

**Issue by Waxpple** [ucb-bar/hammer#650]
Hi,

After I change the config in `example-asap7.yml`
```
...
vlsi.inputs.clocks: [
 {name: "clock_clock", period: "1.25ns", uncertainty: "0.1ns"}
]
```
The design sdc and `sim-syn-input.json` are using 1.25ns setting without problem but the sim-input.yml stays at the 1ns setting.
 Where can I change the `sim-input.yml` to follow the `vlsi.inputs.clocks`?

Thanks!

**Comment by harrisonliew** [comment]
This should be a Chipyard issue, but here's the answer.

`vlsi.inputs.clocks` is the clock period used for synthesis, and gets translated into a timing constraint. In contrast, the clock period in simulation is only dependent on the testbench, which in the case of Chipyard is via the `CLOCK_PERIOD` variable, defined here: https://github.com/ucb-bar/chipyard/blob/main/vcs.mk#L21.

Of course, matching the two clock periods only matters if you are doing timing-annotated gate-level simulations.

**Comment by Waxpple** [comment]
Got it. Thanks!

**Comment by Waxpple** [comment]
Hi,

1. Is there anything else that should be modified if I want to set the clock frequency at 800MHz?
I have changed the `vcs.mk` and `example-asap7.yml`. 
The synthesized design timing is clean (TNS >0) and functional simulation passed.
However, in timing-annotated gate-level simulations, the design failed with timing violation at `frontend.icache`.
And the critical path is not seen in the 
2. I read the issue on (https://github.com/ucb-bar/hammer/issues/644). Soon we can perform post-syn-power analysis? That is great! We can read the `saif` file and get the benchmark power analysis result. I hope I can be part of the PR if it is possible?
 
Thanks.

**Comment by harrisonliew** [comment]
1. Ah, due to the fact that the ASAP7 SRAM models are completely fake, they have no timing annotation for gate-level simulation, and the LIBs have fake timing too. So, there cannot be an expectation for the timing-annotated simulations to pass if there are any dummy SRAMs in the design. In this case, you see a hold violation because in synthesis, the SRAMs have a positive clk-q delay while in simulation, this delay is 0.
2. Yes, the post-syn-power PRs are coming very soon. We'll make sure to tag you for review, even if you don't have Joules in case you want to comment.
3. I have not encountered FGP getting out of sync. It looks like there are some options for FGP such as `sync:busywait`, `sync:mutex`, and `sync:serial`. It looks like it's in decreasing order of performance and increasing order of safety. Can you try these and see what changes? You might also need to combine it with the `diag:ruse` option.

**Comment by Waxpple** [comment]
> 1. Ah, due to the fact that the ASAP7 SRAM models are completely fake, they have no timing annotation for gate-level simulation, and the LIBs have fake timing too. So, there cannot be an expectation for the timing-annotated simulations to pass if there are any dummy SRAMs in the design. In this case, you see a hold violation because in synthesis, the SRAMs have a positive clk-q delay while in simulation, this delay is 0.

I don't understand why would the delay be 0 in the simulation? Don't we have SRAM lib which delay can be calculated and annotated in the simulation?

> 2. Yes, the post-syn-power PRs are coming very soon. We'll make sure to tag you for review, even if you don't have Joules in case you want to comment.

That would be great.

> 3. I have not encountered FGP getting out of sync. It looks like there are some options for FGP such as `sync:busywait`, `sync:mutex`, and `sync:serial`. It looks like it's in decreasing order of performance and increasing order of safety. Can you try these and see what changes? You might also need to combine it with the `diag:ruse` option.

Ok. I will test it out.

**Comment by harrisonliew** [comment]
Oh, I see. Yes, indeed the SDF file encodes the SRAM timing, sorry for not checking this.

You are also correct that Genus can't check hold time, hence you see the violation in Tempus (we are planning to add a Tempus plugin for STA soon). However, Tempus understands the uncertainty (100ps) but VCS does not, so that path in Tempus should not cause a hold violation in VCS...

Another thing to check: what is the effective `timescale` resolution? I'm wondering if the resolution is somehow 100ps instead of the default 10ps (from the Makefiles), hence the setup/hold limits of 100 in the `$setuphold` task? The SDF file indicates the `DFFHQx4` cell should hae a hold time of no more than 14ps...

**Comment by Waxpple** [comment]
> Oh, I see. Yes, indeed the SDF file encodes the SRAM timing, sorry for not checking this.
> 
> You are also correct that Genus can't check hold time, hence you see the violation in Tempus (we are planning to add a Tempus plugin for STA soon). However, Tempus understands the uncertainty (100ps) but VCS does not, so that path in Tempus should not cause a hold violation in VCS...
> 
> Another thing to check: what is the effective `timescale` resolution? I'm wondering if the resolution is somehow 100ps instead of the default 10ps (from the Makefiles), hence the setup/hold limits of 100 in the `$setuphold` task? The SDF file indicates the `DFFHQx4` cell should hae a hold time of no more than 14ps... maybe the resolution should be decreased to 1ps, to match the SDF?

I think both issues exist! 
I override the timescale on vcs compile time and get this:
![image](https://user-images.githubusercontent.com/20642651/167239350-41417a8f-a5f4-4b7a-ada3-b99a2117aa78.png)
However, the sdf annotation failed on SRAM module for some reason.
![image](https://user-images.githubusercontent.com/20642651/167239387-b32c6ef4-335b-4dd0-9f7e-78047d1a2cd3.png)
I am still figuring out why the "timing check annotation is not enable"

**Comment by Waxpple** [comment]
Edit:

Is it because `sram.v` did not use `specify` and the sdf file cannot be back annotated?

**Comment by Waxpple** [comment]
![image](https://user-images.githubusercontent.com/20642651/169059653-16cd8552-2274-47fb-980d-df7fa36cafa2.png)
In the SRAM compiler, we didn't get `bytemask` in our sram Verilog model. And it exists in the LEF file which causes post-syn sim failed

**Comment by Waxpple** [comment]
I add the bytemask for 256x64 and 256x128.
In netlist, `bytemask` signal is connected to `tielo` cell and RTL doesn't even declare the port, so I only add the dummy port for post-syn simulation.

**Comment by harrisonliew** [comment]
I was able to remove BYTEMASK from the LEF/GDS/LIB views for those 2 SRAMs. Please try it out and see if it works. You will need to re-synthesize.

**Comment by harrisonliew** [comment]
> I added the auto-generated specify block in the SRAM compiler.
> The sdf back-annotation is functioning properly now.

Also, can you make a PR with this?

**Comment by Waxpple** [comment]
> > I added the auto-generated specify block in the SRAM compiler.
> > The sdf back-annotation is functioning properly now.
> 
> Also, can you make a PR with this?

Yes, I have made a PR with the patch.
Sorry for the late response, this week or two is our final.

**Comment by harrisonliew** [comment]
Merged, thanks! Good luck with your finals, and hope you can test out #653 afterwards.

**Comment by Waxpple** [comment]
When I run par with floorplan.tcl, the instance name of the SRAMs is not aligned with the design. Where can I change the inst name prefix? I guess the path is fetched from `par-input.json`.

**Comment by harrisonliew** [comment]
The inst name of SRAMs is generated from the MacroCompiler step, and cannot be changed unless you change it in the generated Verilog.

As for the floorplan, you must provide a position for each SRAM instance (see `vlsi.inputs.placement_constraints` in the `example_asap7.yml` file) that you have. You can also edit this key in `par-input.json` directly for debugging first, but this file is generated by Hammer during the flow. Otherwise they will be all be placed outside the design and Innovus will fail.

**Comment by Waxpple** [comment]
In this case, if I change a design from `smallboomconifg` to `mediumboomconfig`. I will have to do the following steps?
1. Check the design SRAMs like which SRAM macro I used and its hierarchical name/path. The used SRAMs can only be traced in the RTL code. (Or maybe there is any JSON I can read directly?)
2. Change the `placement_constraints` in the `example-asap7.yml`.
3. P&R

**Comment by harrisonliew** [comment]
Correct. It is definitely a bit tedious. As for placement guidelines, you should cluster ICache and DCache, and within those clusters, the tag and data arrays should be placed together. Placement/orientation needs to leave room for routing and cannot block pin access. We are working on a research project to automatically generate placement constraints from Chisel.

Ack - good catch on the coreSite. With the ASAP7 r1.7 update, they changed the site to asap7sc7p5t. I'll PR an update for the SRAM LEFs.

**Comment by Waxpple** [comment]
I just use genus to dump the SRAM list from ``get_db insts -if {.is_macro==true}] > sram.log.
There are 316 instances in `SmallBoomConfig`, so should I manually type every macro placement constraint?
It is not easy to know whether the SRAM placement is following DRC. 
I know there is an open-source framework (https://github.com/google-research/circuit_training), which uses the RL model to place macro or std cells. I haven't dived in yet. Maybe It can be used in SRAM placement?

**Comment by harrisonliew** [comment]
You could also just have Innovus do auto-floorplanning with the key `par.innovus.floorplan_mode: auto` (implementation [here](https://github.com/ucb-bar/hammer-cadence-plugins/blob/master/par/innovus/__init__.py#L878)). These will not be DRC clean, however, since the `plan_design` doesn't place it on the valid grid.

Yes, we've also been interested in Google's placer, but we haven't had the resources to play with it.

**Comment by Waxpple** [comment]
> Ack - good catch on the coreSite. With the ASAP7 r1.7 update, they changed the site to asap7sc7p5t. I'll update the https://github.com/ucb-bar/hammer/pull/653 PR with this fix. Edit: done, please test!

I will test this PR today and let you know the result.

> You could also just have Innovus do auto-floorplanning with the key par.innovus.floorplan_mode: auto (implementation [here](https://github.com/ucb-bar/hammer-cadence-plugins/blob/master/par/innovus/__init__.py#L878) - it ignores all your floorplan constraints). These will not be DRC clean, however, since the plan_design doesn't place it on the valid grid.

Nice, I will also test this out.

> Yes, we've also been interested in Google's placer, but we haven't had the resources to play with it.

I think it will be a very nice topic/feature! About resources, do you mean not enough researchers to play with it?

**Comment by harrisonliew** [comment]
> I will test this PR today and let you know the result.

Thanks!

> About resources, do you mean not enough researchers to play with it?

Correct. But also, there are some claims that this doesn't actually improve upon the auto-placers in commercial tools, so that's an entire research project in itself to analyze its performance in generalized designs.

**Comment by Waxpple** [comment]
I am trying to run Genus physical flow. Have you tried this flow before? 
I am still trying to figure this flow out.
If I directly add only `predict_floorplan` before `syn_generic -physical`, It will crash.
So what I am trying now is setting top and bottom routing layer and read def from innovus `plan design` in hammer par.tcl before `syn_generic -phy`. Hope this would work.

If we can get some physical information at the synthesis stage before really run PAR, it would be useful.

**Comment by harrisonliew** [comment]
No, we have not yet, but is something we do want to support.

I'm not familiar with how it works, but can vlsi.inputs.placement_constraints be used to constrain the initial floorplan?

**Comment by Waxpple** [comment]
Yes, you can manually place the hard macro and let genus optimize the data path.
I got the congestion score and its hot-spot score.
After I figure this out, I will create a PR.
![image](https://user-images.githubusercontent.com/20642651/174248998-0dcb39a3-ca22-442f-80d5-24a292beda65.png)

*I guess crashing when using the entire physical flow in genus is the tool bugs at version 20.10.001 or earlier versions.
But there is some workaround.

**Comment by Waxpple** [comment]
1. I am verifying the design, but I still encounter the VCS non-stop issue in `sim-syn-timing-debug`.  
If I use keyboard interrupt (ctrl + c), the simulation will pass. 
If I use max-cycle to time out the simulation, the simulation will fail.
What could cause this weird behavior?

2. I confirmed that GENUS 20.10.001 is buggy with the physical flow, GENUS 21.12 runs the same flow without any problem.

**Comment by Waxpple** [comment]
1. Originally, I use FGP in compilation time and did not use it in run-time. After I remove the FGP in compilation time, the simulation can stop itself with this failed message.
Not sure what does `(tohost = 5)` mean.
![image](https://user-images.githubusercontent.com/20642651/175189096-44836d55-8f73-46b7-b9b9-d58f1515bce8.png)
*The timing violation at the reset stage can be ignored since we reset our design for a long period. This reset stage timing violation will cause python subprocess thinks it is finish like this:
![image](https://user-images.githubusercontent.com/20642651/175189362-21ae621b-d2ad-4df3-a84a-0651432b6419.png)

2. Yes, after I verify the flow and I will make a PR!

**Comment by Waxpple** [comment]
I saw a lot of people encountered this `tohost=????`. https://github.com/ucb-bar/chipyard/issues/753
But I don't know the solution yet. 
I adjust the precision of the simulation timescale from `1ns/1ps` to `1ns/100ps`. The `tohost` problem is gone.

**Comment by Waxpple** [comment]
I have a question, I am still solving the simulation failed with SDF annotation. 
I try to define input/output delay (25% clock period) constraints for synthesis. It doesn't work.
I cannot see any setup violation during synthesis but the simulation failed with timing violations.
Any suggestion?
It will fail even if I use 1ns for synthesis constraints and run the simulation with a 2ns clock period.
